src/common/snippets/src/pass/common_optimizations.cpp

```diff
@@ -68,82 +82,87 @@ void CommonOptimizations::SplitDimensionM(const std::shared_ptr<ov::snippets::op
     // For example, there are 16 threads and shape [6, 512, 32]
     //              LCM(6, 16) = 48 <- ideal work amount for parallelism
     //              new_shape [6, 48 / 6, 512 / (48 / 6), 32 ] => [6, 8, 64, 32]
     //              Each thread has parallelism_work_amount = 6 * 8 / nthrs = 3
-    auto get_lcm = [](size_t a, size_t b) {
-        std::function<size_t(size_t, size_t)> get_gcd;
-        get_gcd = [&get_gcd](size_t a, size_t b) {
-            if (b == 0)
-                return a;
-            return get_gcd(b, a % b);
-        };
-        return a / get_gcd(a, b) * b;
-    };
     const auto lcm = get_lcm(batch_dim, optimal_parallelism_work_amount);  // LCM(b, nthrs)
     const auto batch_dim_multiplier = lcm / batch_dim;  // LCM(b, nthrs) / b
     const auto needed_new_dim = m_dim / batch_dim_multiplier;  // m / (LCM(b, nthrs) / b) - needed factors of dimension m
-
-    auto is_optimized = [&](size_t batch_m_dim, size_t new_m_dim) {
-        return batch_m_dim != 1 && new_m_dim >= static_cast<size_t>(optimal_m_dim);
-    };
-
     if (batch_dim_multiplier * needed_new_dim == m_dim) {
         batch_m_dim = batch_dim_multiplier;
         new_m_dim = needed_new_dim;
     }
+
+    // [ Second Step ]
+    // If we couldn't optimally split on the previous step, try the second step.
+    // The algorithm finds the more optimal parallelism work amount [batch_dim * batch_m_dim],
+    // where batch_m_dim is divisor of dimension M.
+    // The optimal parallelism work amount means the case when as many threads as possible have work
+    // For example, there are 8 threads and shape [5, 384, 32]
+    //              768 = [2 x 192] = [3 x 128] = [4 x 96] = [6 x 64]
+    //               - [5, 2, 192, 32] - WA = 10 = 8 + 2 (6 threads calculates once and 2 threads twice)
+    //               - [5, 3, 128, 32] - WA = 15 = 8 + 7 (all threads have 2 kernel except one thread) <- the most optimal case
+    //               - [5, 4,  96, 32] - WA = 20 = 8 x 2 + 4
+    //               - [5, 6,  64, 32] - WA = 30 = 8 x 3 + 6
+    // The most optimal and possible case is [5, 3, 128, 32] - almost all threads executes kernel twice
     if (!is_optimized(batch_m_dim, new_m_dim)) {
-        auto get_factors = [](size_t dim) -> std::vector<size_t> {
-            std::vector<size_t> factors;
-            size_t div = 2;
-            while (div <= dim) {
-                const auto res = dim / div;
-                if (res * div == dim) {
-                    factors.push_back(div);
-                    dim = res;
-                } else {
-                    div++;
-                }
+        batch_m_dim = 1;
+        new_m_dim = m_dim;
+        // Heuristic value for a quick exit from the algorithm.
+        // The value shows the number of threads in percentages that perform the most equal work
+        const auto optimal_thread_num_percent = 0.8;
+        size_t opt_remainder = batch_m_dim % optimal_parallelism_work_amount;
```

Comment:
`opt_remainder` always equals to `1`, since `batch_m_dim` is 1. Can we just assign 1?

Reply:
Oh, great catch! Thanks a lot! There should be `size_t opt_remainder = (batch * batch_m_dim) % optimal_parallelism_work_amount;` (like on 145 line) because `opt_remainder` is just the maximum of possible remainders. Firstly we have shape `[batch, batch_m_dim, new_m_dim, smth]` where `batch_m_dim = 1`. So at the beginning we have parallel_work_amount `[batch, 1]`. It means that only `(batch * 1) % optimal_parallelism_work_amount` threads will work. After that we try to increase `batch_m_dim` and calculate new `remainder` as `batch_dim * divisor_0 % optimal_parallelism_work_amount` (`divisor_0` is possible new `batch_m_dim`). If now the current `remainder` is bigger than `opt_remainder`, we update `opt_remainder` and save `divisor_0` and `divisor_1` since they're potential  `batch_m_dim` and `new_m_dim`.
Thanks!